src/main/java/org/folio/consortia/messaging/topic/KafkaTopicsInitializer.java

```diff
@@ -0,0 +1,84 @@
+package org.folio.consortia.messaging.topic;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.apache.kafka.clients.admin.NewTopic;
+import org.folio.consortia.config.kafka.properties.FolioKafkaProperties;
+import org.folio.consortia.messaging.domain.ConsortiaInputEventTypes;
+import org.folio.consortia.messaging.domain.ConsortiaOutputEventTypes;
+import org.folio.spring.FolioExecutionContext;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.kafka.config.KafkaListenerEndpointRegistry;
+import org.springframework.kafka.config.TopicBuilder;
+import org.springframework.kafka.core.KafkaAdmin;
+import org.springframework.stereotype.Component;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Component
+@RequiredArgsConstructor
+public class KafkaTopicsInitializer {
+
+  private final KafkaAdmin kafkaAdmin;
+  private final BeanFactory beanFactory;
+  private final FolioExecutionContext folioExecutionContext;
+  private final KafkaListenerEndpointRegistry kafkaListenerEndpointRegistry;
+  private final FolioKafkaProperties folioKafkaProperties;
+  private final String kafkaEnvId;
+
+  public void createTopics() {
+    if (folioExecutionContext == null) {
+      throw new IllegalStateException("Could be executed only in Folio-request scope");
+    }
+    var tenantId = folioExecutionContext.getTenantId();
+    var topicList = tenantSpecificTopics(tenantId);
+
+    log.info("Creating topics for kafka [topics: {}]", topicList);
+    var configurableBeanFactory = (ConfigurableBeanFactory) beanFactory;
+    topicList.forEach(newTopic -> {
+      var beanName = newTopic.name() + ".topic";
+      if (!configurableBeanFactory.containsBean(beanName)) {
+        configurableBeanFactory.registerSingleton(beanName, newTopic);
+      }
+    });
+    kafkaAdmin.initialize();
```

Comment:
We have very similar logic in KafkaService, can it be combined?

Reply:
cleaned up